rioxarray/rioxarray.py

```diff
@@ -121,23 +121,39 @@ def affine_to_coords(
 
 def _generate_spatial_coords(
     *, affine: Affine, width: int, height: int
-) -> dict[Hashable, Any]:
+) -> xarray.Coordinates:
     """get spatial coords in new transform"""
-    new_spatial_coords = affine_to_coords(affine, width, height)
-    if new_spatial_coords["x"].ndim == 1:
-        return {
-            "x": xarray.IndexVariable("x", new_spatial_coords["x"]),
-            "y": xarray.IndexVariable("y", new_spatial_coords["y"]),
-        }
-    return {
-        "xc": (("y", "x"), new_spatial_coords["x"]),
-        "yc": (("y", "x"), new_spatial_coords["y"]),
-    }
+
+    if get_option("use_raster_index"):
+        try:
+            from rasterix import RasterIndex
+        except (ImportError, ModuleNotFoundError) as err:
+            raise ImportError(
+                "The rasterix package is needed for generating "
+                "the spatial coordinates with a RasterIndex"
+            ) from err
+
+        raster_index = RasterIndex.from_transform(affine, width, height)
```

Comment:
I think we should also delete the transform from the grid_mapping variable like so:
```python
grid_mapping = ds.rio.grid_mapping
obj[grid_mapping].attrs.pop("GeoTransform")
```

following https://github.com/dcherian/rasterix/blob/main/design_notes/raster_index.md#read-time and our discussions. This will trigger some more changes (`_cached_transform` will need to pull the Affine from RasterIndex)